File: backend/routes/career_switch.py
"""
API routes for career switch intelligence
"""
from fastapi import APIRouter, HTTPException, status, Query, Body
from models.schemas import BaseResponse, ErrorResponse
from services.career_switch_service import CareerSwitchService
from typing import Optional
from pydantic import BaseModel, Field

router = APIRouter()
switch_service = CareerSwitchService()


class CareerSwitchRequest(BaseModel):
    """Request schema for career switch analysis"""
    source_career_id: str = Field(..., description="Career ID of current occupation")
    target_career_id: str = Field(..., description="Career ID of target occupation")


class CareerSwitchByNameRequest(BaseModel):
    """Request schema for career switch analysis using career names directly"""
    source_career_name: str = Field(..., description="Name of current career (as typed by user)")
    target_career_name: str = Field(..., description="Name of target career (as typed by user)")


@router.post("/analyze", response_model=BaseResponse)
async def analyze_career_switch(request: CareerSwitchRequest = Body(...)):
    """
    Analyze a career switch from source to target occupation
    
    Returns skill overlap, transfer map, difficulty classification, 
    transition time estimate, and success/risk factors
    """
    try:
        result = switch_service.analyze_career_switch(
            source_career_id=request.source_career_id,
            target_career_id=request.target_career_id
        )
        
        if "error" in result:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=ErrorResponse(
                    success=False,
                    message=result["error"]
                ).model_dump()
            )
        
        return BaseResponse(
            success=True,
            message="Career switch analysis completed",
            data=result
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=ErrorResponse(
                success=False,
                message="Failed to analyze career switch",
                error=str(e)
            ).model_dump()
        )


@router.post("/switch", response_model=BaseResponse)
async def career_switch(request: CareerSwitchRequest = Body(...)):
    """
    Analyze a career switch from source to target occupation.
    
    Returns:
    - overlap %: Percentage of skills that overlap between source and target
    - difficulty: Transition difficulty classification (Low, Medium, High)
    - transition time range: Estimated time range for the transition
    - skill translation map: Structured map of skills (transfers directly, needs learning, optional)
    - success factors + risk factors: Non-deterministic assessment of transition factors
    """
    try:
        result = switch_service.analyze_career_switch(
            source_career_id=request.source_career_id,
            target_career_id=request.target_career_id
        )
        
        if "error" in result:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=ErrorResponse(
                    success=False,
                    message=result["error"]
                ).model_dump()
            )
        
        # Format response with requested fields
        formatted_result = {
            "overlap_percentage": result["skill_overlap"]["percentage"],
            "difficulty": result["difficulty"],
            "transition_time_range": {
                "min_months": result["transition_time"]["min_months"],
                "max_months": result["transition_time"]["max_months"],
                "range": result["transition_time"]["range"],
                "note": result["transition_time"].get("note", "")
            },
            "skill_translation_map": {
                "transfers_directly": result["transfer_map"]["transfers_directly"],
                "needs_learning": result["transfer_map"]["needs_learning"],
                "optional_skills": result["transfer_map"]["optional_skills"]
            },
            "success_factors": result["success_risk_assessment"]["success_factors"],
            "risk_factors": result["success_risk_assessment"]["risk_factors"],
            "overall_assessment": result["success_risk_assessment"].get("overall_assessment", ""),
            "source_career": result["source_career"],
            "target_career": result["target_career"]
        }
        
        return BaseResponse(
            success=True,
            message="Career switch analysis completed",
            data=formatted_result
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=ErrorResponse(
                success=False,
                message="Failed to analyze career switch",
                error=str(e)
            ).model_dump()
        )


@router.post("/switch-by-name", response_model=BaseResponse)
async def career_switch_by_name(request: CareerSwitchByNameRequest = Body(...)):
    """
    Analyze a career switch using career names directly (no database lookup required).
    Uses OpenAI to generate analysis based on the career names provided.
    """
    try:
        result = switch_service.analyze_career_switch_by_name(
            source_career_name=request.source_career_name,
            target_career_name=request.target_career_name
        )
        
        if "error" in result:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=ErrorResponse(
                    success=False,
                    message=result["error"]
                ).model_dump()
            )
        
        # Format response with requested fields
        formatted_result = {
            "overlap_percentage": result.get("overlap_percentage", 0),
            "difficulty": result.get("difficulty", "Medium"),
            "transition_time_range": result.get("transition_time_range", {
                "min_months": 6,
                "max_months": 12,
                "range": "6-12 months",
                "note": "Estimated transition time"
            }),
            "skill_translation_map": result.get("skill_translation_map", {
                "transfers_directly": [],
                "needs_learning": [],
                "optional_skills": []
            }),
            "success_factors": result.get("success_factors", []),
            "risk_factors": result.get("risk_factors", []),
            "overall_assessment": result.get("overall_assessment", ""),
            "source_career": {
                "career_id": result.get("source_career", {}).get("career_id", ""),
                "name": request.source_career_name
            },
            "target_career": {
                "career_id": result.get("target_career", {}).get("career_id", ""),
                "name": request.target_career_name
            }
        }
        
        return BaseResponse(
            success=True,
            message="Career switch analysis completed",
            data=formatted_result
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=ErrorResponse(
                success=False,
                message="Failed to analyze career switch",
                error=str(e)
            ).model_dump()
        )


@router.get("/overlap", response_model=BaseResponse)
async def get_skill_overlap(
    source: str = Query(..., description="Source career ID"),
    target: str = Query(..., description="Target career ID")
):
    """
    Get skill overlap percentage and transfer map between two careers
    Quick endpoint for just the overlap data
    """
    try:
        result = switch_service.compute_skill_overlap(source, target)
        
        if "error" in result:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=ErrorResponse(
                    success=False,
                    message=result["error"]
                ).model_dump()
            )
        
        return BaseResponse(
            success=True,
            message="Skill overlap computed",
            data=result
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=ErrorResponse(
                success=False,
                message="Failed to compute skill overlap",
                error=str(e)
            ).model_dump()
        )

File: frontend/src/services/careerSwitch.ts
/**
 * Career Switch Service
 * Service for analyzing career transitions and skill overlaps
 */
import apiClient from './apiClient';
import { BaseResponse } from './types';

/**
 * Career switch request
 */
export interface CareerSwitchRequest {
  source_career_id: string;
  target_career_id: string;
}

/**
 * Career switch by name request
 */
export interface CareerSwitchByNameRequest {
  source_career_name: string;
  target_career_name: string;
}

/**
 * Skill item structure
 */
export interface SkillItem {
  skill: string;
  source_level: number;
  target_level: number;
  gap?: number; // Only present in needs_learning
}

/**
 * Skill translation map
 */
export interface SkillTranslationMap {
  transfers_directly: SkillItem[];
  needs_learning: SkillItem[];
  optional_skills: SkillItem[];
}

/**
 * Success and risk factors
 */
export interface SuccessRiskAssessment {
  success_factors: string[];
  risk_factors: string[];
  overall_assessment?: string;
}

/**
 * Career switch analysis response (from /analyze endpoint)
 */
export interface CareerSwitchAnalysis {
  skill_overlap: {
    percentage: number;
    similarity_score: number;
  };
  transfer_map: SkillTranslationMap;
  difficulty: 'Low' | 'Medium' | 'High';
  transition_time: {
    min_months: number;
    max_months: number;
    range: string;
    note?: string;
  };
  success_risk_assessment: SuccessRiskAssessment;
  source_career: {
    career_id: string;
    name: string;
    soc_code: string;
  };
  target_career: {
    career_id: string;
    name: string;
    soc_code: string;
  };
}

/**
 * Factor item structure
 */
export interface FactorItem {
  factor: string;
  description: string;
  impact: 'positive' | 'negative';
}

/**
 * Formatted career switch response (from /switch endpoint)
 */
export interface CareerSwitchFormatted {
  overlap_percentage: number;
  difficulty: 'Low' | 'Medium' | 'High';
  transition_time_range: {
    min_months: number;
    max_months: number;
    range: string;
    note?: string;
  };
  skill_translation_map: SkillTranslationMap;
  success_factors: FactorItem[];
  risk_factors: FactorItem[];
  overall_assessment?: string;
  source_career: {
    career_id: string;
    name: string;
    soc_code: string;
  };
  target_career: {
    career_id: string;
    name: string;
    soc_code: string;
  };
}

/**
 * Skill overlap response (from /overlap endpoint)
 */
export interface SkillOverlap {
  overlap_percentage: number;
  similarity_score: number;
  transfer_map: SkillTranslationMap;
  source_career: {
    career_id: string;
    name: string;
  };
  target_career: {
    career_id: string;
    name: string;
  };
}

/**
 * POST /api/career-switch/analyze - Analyze career switch
 * Returns detailed analysis of career transition including skill overlap, transfer map, difficulty, and time estimates
 */
export const analyzeCareerSwitch = async (
  request: CareerSwitchRequest
): Promise<CareerSwitchAnalysis> => {
  const response = await apiClient.post<BaseResponse<CareerSwitchAnalysis>>(
    '/api/career-switch/analyze',
    request
  );

  if (!response.data.success) {
    throw new Error(response.data.message || 'Failed to analyze career switch');
  }

  return response.data.data!;
};

/**
 * POST /api/career-switch/switch - Career switch (formatted response)
 * Returns formatted career switch analysis with overlap percentage, difficulty, transition time, and skill translation map
 */
export const getCareerSwitch = async (
  request: CareerSwitchRequest
): Promise<CareerSwitchFormatted> => {
  const response = await apiClient.post<BaseResponse<CareerSwitchFormatted>>(
    '/api/career-switch/switch',
    request
  );

  if (!response.data.success) {
    throw new Error(response.data.message || 'Failed to get career switch analysis');
  }

  return response.data.data!;
};

/**
 * POST /api/career-switch/switch-by-name - Career switch using career names directly
 * Uses OpenAI to generate analysis based on the career names provided (no database lookup)
 */
export const getCareerSwitchByName = async (
  request: CareerSwitchByNameRequest
): Promise<CareerSwitchFormatted> => {
  const response = await apiClient.post<BaseResponse<CareerSwitchFormatted>>(
    '/api/career-switch/switch-by-name',
    request
  );

  if (!response.data.success) {
    throw new Error(response.data.message || 'Failed to get career switch analysis');
  }

  return response.data.data!;
};

/**
 * GET /api/career-switch/overlap - Get skill overlap (query params)
 * Quick endpoint for just the overlap data between two careers
 */
export const getSkillOverlap = async (
  source: string,
  target: string
): Promise<SkillOverlap> => {
  const response = await apiClient.get<BaseResponse<SkillOverlap>>(
    '/api/career-switch/overlap',
    {
      params: {
        source,
        target,
      },
    }
  );

  if (!response.data.success) {
    throw new Error(response.data.message || 'Failed to get skill overlap');
  }

  return response.data.data!;
};

/**
 * Career switch service object with all methods
 */
export const careerSwitchService = {
  analyzeCareerSwitch,
  getCareerSwitch,
  getCareerSwitchByName,
  getSkillOverlap,
};

export default careerSwitchService;

